rskj-core/src/main/java/org/ethereum/listener/GasPriceTracker.java

```diff
@@ -61,30 +75,39 @@ public void onBlock(Block block, List<TransactionReceipt> receipts) {
         logger.trace("End onBlock");
     }
 
-    public void onTransaction(Transaction tx) {
+    private void onTransaction(Transaction tx) {
         if (tx instanceof RemascTransaction) {
             return;
         }
 
         if (idx == -1) {
-            idx = window.length - 1;
-            filled = true;
+            idx = WINDOW_SIZE - 1;
             lastVal = null;  // recalculate only 'sometimes'
         }
 
         window[idx--] = tx.getGasPrice();
     }
 
-    public Coin getGasPrice() {
-        if (!filled) {
+    public synchronized Coin getGasPrice() {
+        if (window[0] == null) { // not filled yet
             return defaultPrice;
         } else {
             if (lastVal == null) {
-                Coin[] values = Arrays.copyOf(window, window.length);
+                Coin[] values = Arrays.copyOf(window, WINDOW_SIZE);
                 Arrays.sort(values);
                 lastVal = values[values.length / 4];  // 25% percentile
             }
-            return lastVal;
+
+            Coin bestBlockPrice = bestBlockPriceRef.get();
+            if (bestBlockPrice == null) {
+                return lastVal;
+            } else {
+                return max(lastVal, bestBlockPrice.multiply(BI_11).divide(BI_10));
+            }
         }
     }
+
+    private static Coin max(Coin a, Coin b) {
```

Comment:
wouldnt be better to declare max() in Coin? would be consistent with BigInteger.max()

Reply:
nice :)